### DISCUSSION TITLE
Integration with cypress

### DISCUSSION THREAD

**Discussion by mammadataei** [testing-library/jest-dom#445] [Q&A]
Hi, Does anyone have a solution to integrate `jest-dom` with cypress? or any alternative? I need some of the assertions like `toHaveAccessibleDescription`, `toHaveAccessibleName` and `toHaveErrorMessage` which are not available in cypress or `chai-dom`.

**Comment by gnapse** [comment]
I'd love to have something like that.

It'd be great if we could export lower level functions that implement the logic of the checks that each assertion does. Then another library (perhaps a @testing-library/cypress-dom` library) could then implement Cypress-centric chai-like assertions while re-using the same implementation.

**Reply by mammadataei** [reply]
I agree, It would be great to have a flexible foundation for integrating with different tools and libraries.

**Reply by mammadataei** [reply]
Okay, I had some experiments with cypress and chai and wrote a couple of `jest-dom` assertions for chai. I think I can contribute.

**Reply by gnapse** [reply]
Looking forward to your contribution. Care to share/brainstorm how you plan to do it? How are you planning to "extract" the core assertion implementation in a way that can still power the jest-dom jest-centric matchers while also allowing others to plug them in somewhere else?

**Reply by mammadataei** [reply]
Well, It might be a good use case for a monorepo with tree packages. One for core, one for jest, and the other one for chai/cypress. But it depends if we can change the `jest-dom` repo.

I think the core package should contain functions for performing checks and calculations for each assertion which return a boolean indicating if the assertion is passing or not.
Then, consumer packages like `jest-dom` only take care of generating the proper error message for the user.

Cononsider the [`toBeInvalid`](https://github.com/testing-library/jest-dom/blob/main/src/to-be-invalid.js) as an example:

```javascript
import { isElementInvalid } from '@testing-library/assertion-core' 

// jest
export function toBeInvalid(element) {
  const isInvalid = isElementInvalid(element)

  return {
    pass: isInvalid,
    message: () => {
      const is = isInvalid ? 'is' : 'is not'
      return [
        this.utils.matcherHint(
          `${this.isNot ? '.not' : ''}.toBeInvalid`,
          'element',
          '',
        ),
        '',
        `Received element ${is} currently invalid:`,
        `  ${this.utils.printReceived(element.cloneNode(false))}`,
      ].join('\n')
    },
  }
}

// chai
_chai.Assertion.addMethod('invalid', function (expected) {
    const element = utils.flag(this, 'object').get(0)
    const isInvalid = isElementInvalid(element)

    this.assert(
      isInvalid,
      'expected the element to be invalid.',
      'expected the element not to be invalid.',
      element,
    )
  })
```

**Reply by gnapse** [reply]
I love the idea. I do think this is the right direction.

Although I'm hesitant of the prospect of having to manage a monorepo (mostly because I have no experience dealing with it). One thing I would not want to lose is the automated releases system provided by `kcd-scripts`.

Also, this being a non-trivial architectural change of how it all works, I want to involve @testing-library/core-maintainers to see what they think.

**Reply by mammadataei** [reply]
I understand. Monorepo always requires some considerations. In this case, I thought it might be easier to manage the packages because they will be dependent on each other.

But It's not a big deal and we can continue with the multi-repo approach like other testing-library packages.

And I think it's a good idea to develop the `core` and `chai` packages to make sure everything is fine, and then start to use `core` for refactoring the `jest-dom`.

**Reply by kentcdodds** [reply]
Would be really cool to be able to use these assertions in Cypress. I don't have any time to consult or work on it myself, but I'm definitely for it if it doesn't put undue stress on the codebase.

**Reply by alexkrolick** [reply]
+1 to this. I think going the Chai route makes sense since it's a standard extension system.

**Reply by mammadataei** [reply]
Okay, if we're all on the same page, how can I begin to contribute?

**Reply by gnapse** [reply]
Do you think a new repo is the way to go? On one hand, I'm hesitant to adopt monorepos. On the other hand, having the assertions implements in a different package will create some indirection when fixing bugs in jest-dom (having to fix and release them in another package first, then upgrade the dependency in jest-dom). But I guess there's no way around it, one way or the other.

Anyway, I can create a new repository called `@testing-library/dom-assertions` and we can start extracting into it the core implementation of all the assertions here. Once we have them there, we can work on stripping this repo of all that code and have the assertions use the functions exported from his new package. We can keep the tests in jest-dom and run them with the new implementation to make sure nothing broke.

Sounds like a plan?

**Reply by mammadataei** [reply]
Yeah, sounds like a plan.

**Reply by mammadataei** [reply]
Hi, 
@gnapse any news about repo?

**Reply by eps1lon** [reply]
@mammadataei You could start the work in your account and we changer ownership of the repo later. So create it in https://github.com/mammadataei/dom-assertions and later we transfer it to https://github.com/@testing-library/dom-assertions

**Reply by mammadataei** [reply]
Good. just one more thing to ask. Does the project's setup or toolchain matter? Or Are there any conventions I have to follow?

**Reply by gnapse** [reply]
Thanks, @eps1lon for that reply. Indeed, that's a very good approach.

> Does the project's setup or toolchain matter? Or Are there any conventions I have to follow?

It'd be good if you base your initial repo structure off of copying the way jest-dom is set up. In particular, the way the package json scripts section works, which is mostly based on kcd-scripts, and how it does versioning. This will allow the automated semantic versioning to work, and make it easier to automate releases on PR merges.

**Reply by mammadataei** [reply]
Okay, I'll try to set up the structure and the scripts as is in `jest-dom`, but do you mind if I use `typescript`, `vitest`, and probably `pnpm` (for better developer experience)?

**Reply by gnapse** [reply]
About TypeScript, definitely.

About vitest and pnpm, I have never used them, so I may be negatively biased here, but I would not want this package to divert too much from the developer experience in all the other packages in testing-library. It is my understanding that we mostly use jest and npm. But feel free to make your case here if you feel strongly about it. Or let's see what others think.

**Reply by mammadataei** [reply]
@gnapse @eps1lon @kentcdodds 

Hey again, It's been quite a while 😁, and eventually, I got some time and wrote a couple of assertions. Here's the [repo](https://github.com/mammadataei/dom-assertions). There ain't much but I think it's enough for an alpha release to test if the API integrates properly with `jest-dom`. At the same time, I can continue adding more assertions and start `cypress-dom` as well.
